libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,923 @@
+use std::{ str::Chars, iter::Peekable, collections::HashMap, slice::Iter };
+
+use crate::semantic::types::*;
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum TokenTypes {
+    ERROR,
+    EOF, 
+
+    // Keywords and Kinds
+    FORALL,
+    WHERE,
+    IS,
+    AND, 
+
+    // Literals and Indentifiers
+    INT,
+    FLOAT,
+    STRING,
+    BOOL,
+    REGEX,
+    IDENTIFIER,
+    WHITESPACE,
+
+    // Operators
+    LEFTCURLYBRAC,
+    RIGHTCURLYBRAC,
+    LEFTSQUAREBRAC,
+    RIGHTSQUAREBRAC,
+    LEFTPAREN,
+    RIGHTPAREN,
+    QUESTIONMARK,
+    COLON, 
+    COMMA, 
+    OR, 
+    PIPE, 
+    ARROW, 
+}
+
+
+struct Lexer<'a> {
+    source: Peekable<Chars<'a>>, 
+    tokens: Vec<Token>, 
+    current_string: String,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Token {
+    token_type: TokenTypes,
+    text: Option<String>,   
+}
+
+impl Lexer<'_> {
+    fn next(&mut self) -> Option<char> {        
+        match self.source.next() {
+            None => None, 
+            Some(letter) => {
+                // push current letter into current_string
+                if letter.is_alphanumeric() {
+                    self.current_string.push(letter);
+                }
+                Some(letter)
+            }
+        }
+    }
+
+    fn emit(&mut self, token: TokenTypes) {
+        if !self.current_string.is_empty() {
+            let text = self.current_string.clone();
+            // if the token requires a string, it should always be in self.current_string
+            self.tokens.push(Token{ token_type: token, text: Some(text) })
+        } else {
+            self.tokens.push(Token{ token_type: token, text: None })
+        }
```

Comment:
At least one case comes to mind where the parser logic checks to see if the next token has text or not. I think the logic could be changed, but I'd need to look into it a little more. 

As far as simplifying the logic, I'm not sure if this change would have a big impact. I already unpack the text in a helper function, so from the perspective of anyone using the Tokens as is, this seems like it would be minor. But perhaps it might cost less in terms of memory to just use an empty string rather than creating an `Option`.

Reply:
For that initial case that I mentioned, I could also check if `String.is_empty()` if I change this per your suggestion.

Reply:
As as example, the `parse_type_vars` function uses the text Option.